clients/src/test/java/org/apache/kafka/test/TestUtils.java

```diff
@@ -353,28 +355,77 @@ public static void waitForCondition(final TestCondition testCondition, final Sup
+     * @throws InterruptedException if the current thread is interrupted while waiting for {@code runnable} to complete successfully.
+     */
+    public static void retryOnExceptionWithTimeout(final long timeoutMs,
+                                                   final ValuelessCallable runnable) throws InterruptedException {
+        retryOnExceptionWithTimeout(DEFAULT_POLL_INTERVAL_MS, timeoutMs, runnable);
+    }
 
-        // don't re-evaluate testCondition.conditionMet() because this might slow down some tests significantly (this
-        // could be avoided by making the implementations more robust, but we have a large number of such implementations
-        // and it's easier to simply avoid the issue altogether)
-        if (!testConditionMet) {
-            String conditionDetailsSupplied = conditionDetailsSupplier != null ? conditionDetailsSupplier.get() : null;
-            String conditionDetails = conditionDetailsSupplied != null ? conditionDetailsSupplied : "";
-            throw new AssertionError("Condition not met within timeout " + maxWaitMs + ". " + conditionDetails);
+    /**
+     * Wait for the given runnable to complete successfully, i.e. throw now {@link Exception}s or
+     * {@link AssertionError}s, or for the default timeout to expire. If the timeout expires then the
+     * last exception or assertion failure will be thrown thus providing context for the failure.
+     *
+     * @param runnable the code to attempt to execute successfully.
+     * @throws InterruptedException if the current thread is interrupted while waiting for {@code runnable} to complete successfully.
+     */
+    public static void retryOnExceptionWithTimeout(final ValuelessCallable runnable) throws InterruptedException {
+        retryOnExceptionWithTimeout(DEFAULT_POLL_INTERVAL_MS, DEFAULT_MAX_WAIT_MS, runnable);
+    }
+
+    /**
+     * Wait for the given runnable to complete successfully, i.e. throw now {@link Exception}s or
+     * {@link AssertionError}s, or for the given timeout to expire. If the timeout expires then the
+     * last exception or assertion failure will be thrown thus providing context for the failure.
+     *
+     * @param pollIntervalMs the interval in milliseconds to wait between invoking {@code runnable}.
+     * @param timeoutMs the total time in milliseconds to wait for {@code runnable} to complete successfully.
+     * @param runnable the code to attempt to execute successfully.
+     * @throws InterruptedException if the current thread is interrupted while waiting for {@code runnable} to complete successfully.
+     */
+    public static void retryOnExceptionWithTimeout(final long pollIntervalMs,
+                                                   final long timeoutMs,
+                                                   final ValuelessCallable runnable) throws InterruptedException {
+        final long expectedEnd = System.currentTimeMillis() + timeoutMs;
+
+        while (true) {
```

Comment:
IMHO it seems we may be susceptible to a hanging test here.  Should we change the condition of the `while` loop from to have a check that the `expectedEnd ` has elapsed?

Reply:
I don't want to put the condition in the while condition expression because I want to be sure the test runs at least once. I could add a separate boolean to check if we've run once, but don't want to overcomplicate it if not needed. I'm not sure why there is a risk of hanging?

We always hit runnable.call. Assuming it doesn't hang itself (which we can't do anything about, short of spinning up another thread), it will do one of:

1. Return successfully, in which case the whole function returns.
2. Throws an Exception, in which case we test if the timeout has expired and a) rethrow if it has or b) sleep and retest if it has not.
3. Throws an Assertion error, in which case we handled it the same way as 2.
4. Throws an Error, in which case the we also unwind the stack, just without rethrowing.

Let me know if I missed a case. Thanks!

Reply:
Also looking at it again, we would need to keep ahold of the last exception to be able to rethrow it when we exit the while loop - in fact that could act like our boolean. Still, I would prefer to keep this simple if possible. Please do let me know if I missed something in my analysis. Thanks again for the review!

Reply:
Thanks for the detailed explanation, SGTM